### ISSUE TITLE
`nix` >= 2.6.0 will not build on darwin systems

### ISSUE THREAD

**Issue by hexagonal-sun** [NixOS/nixpkgs#157201]
### Describe the bug
As the title says.  Build log attached.

### Steps To Reproduce
`nix build "nixpkgs#nixUnstable"` on an `aarch64-darwin` machine.

### Expected behavior
The package should build.

### Additional context
See attached build log.

[nixUnstable-build-failure.log](https://github.com/NixOS/nixpkgs/files/7962122/nixUnstable-build-failure.log)
### Notify maintainers

@edolstra 
@NixOS/darwin-maintainers 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 20.6.0, macOS 11.6`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.5pre20211007_844dd90`
 - channels(matthew): `"darwin, home-manager, nixpkgs-21.11pre330079.bf346b557b7"`
 - channels(root): `"home-manager, nixpkgs-21.11pre328902.2606cb0fc24"`
 - nixpkgs: `/nix/store/08yiicpdx3y7j6b6g6p1ghipllxbrcy2-pnwyi8fcq1v0b9i4905b3yxkaw29dmly-source`

```

**Comment by jonringer** [comment]
AFAICT, this may be from the c++ version being c++11 for darwin's stdenv. Where as some c++14 usage is occuring:
```
src/libstore/realisation.hh:22:5: error: no matching function for call to 'make_tuple'
    GENERATE_CMP(DrvOutput, me->drvHash, me->outputName);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/libutil/comparator.hh:31:5: note: expanded from macro 'GENERATE_CMP'
    GENERATE_LEQ(args) \
    ^~~~~~~~~~~~~~~~~~
src/libutil/comparator.hh:27:31: note: expanded from macro 'GENERATE_LEQ'
#define GENERATE_LEQ(args...) GENERATE_ONE_CMP(<, args)
                              ^~~~~~~~~~~~~~~~~~~~~~~~~
src/libutil/comparator.hh:23:22: note: expanded from macro 'GENERATE_ONE_CMP'
      auto fields2 = std::make_tuple( FIELDS ); \
```
https://en.cppreference.com/w/cpp/utility/tuple/make_tuple

**Comment by zowoq** [comment]
Likely the same error for `nix` `2.6.0` on both darwin platforms as well.

**Comment by lovesegfault** [comment]
Would forcing the usage of a sufficiently new `llvmPackages` on Darwin be sufficient?

I do not have access to the necessary hardware to test/debug this, someone will need to help.

**Comment by jonringer** [comment]
I believe aarch64 darwin is on a very recent version of llvm

**Comment by hexagonal-sun** [comment]
I've just confirmed that this problem also exists on `x86_64-darwin` too.  I've renamed the issue to reflect this.

**Comment by zowoq** [comment]
> I believe aarch64 darwin is on a very recent version of llvm

Both darwin platforms are on llvm 11.

> I've just confirmed that this problem exists on x86_64-darwin too. I've renamed the issue to reflect this.

I checked and it is the same error for 2.6.0 also so I've renamed the issue again.
_____

https://hydra.nixos.org/project/nix

Apart from some flaky tests the `nix` repo hydra jobset seems to be building >= 2.6.0 and it looks to be using a `nixpkgs` revision from `21.05`?

**Comment by kubukoz** [comment]
FWIW I'm able to build it from the upstream repo's flake but not from nixpkgs

**Comment by hexagonal-sun** [comment]
I've managed to get a fix which is up for review in  #157596.
